src/webui/api/torrentscontroller.cpp

```diff
@@ -1768,3 +1946,193 @@ void TorrentsController::setSSLParametersAction()
+        throw APIError(APIErrorType::BadParams, tr("Must specify URI or hash"));
+
+    const QString source = QUrl::fromPercentEncoding(sourceParam.toLatin1());
+
+    BitTorrent::InfoHash infoHash;
+    if (const auto iter = m_torrentSourceCache.constFind(source); iter != m_torrentSourceCache.constEnd())
+        infoHash = iter.value();
+    else if (const auto sourceTorrentDescr = BitTorrent::TorrentDescriptor::parse(source))
+        infoHash = sourceTorrentDescr.value().infoHash();
+
+    if (!infoHash.isValid())
+        throw APIError(APIErrorType::NotFound);
+
+    const BitTorrent::TorrentDescriptor &torrentDescr = m_torrentMetadataCache[infoHash];
+    if (!torrentDescr.info().has_value())
+        throw APIError(APIErrorType::Conflict, tr("Metadata is not yet available"));
+
+    const nonstd::expected<QByteArray, QString> result = torrentDescr.saveToBuffer();
+    if (!result)
+        throw APIError(APIErrorType::Conflict, tr("Unable to export torrent metadata. Error: %1").arg(result.error()));
+
+    setResult(result.value(), u"application/x-bittorrent"_s, (infoHash.toTorrentID().toString() + u".torrent"));
+}
+
+void TorrentsController::onDownloadFinished(const Net::DownloadResult &result)
+{
+    const QString source = result.url;
+    m_requestedTorrentSource.remove(source);
+
+    switch (result.status)
+    {
+    case Net::DownloadStatus::Success:
+        // use the info directly from the .torrent file
+        if (const auto loadResult = BitTorrent::TorrentDescriptor::load(result.data))
+        {
+            const BitTorrent::TorrentDescriptor &torrentDescr = loadResult.value();
+            const BitTorrent::InfoHash infoHash = torrentDescr.infoHash();
+            m_torrentSourceCache.insert(source, infoHash);
+            m_torrentMetadataCache.insert(infoHash, torrentDescr);
+        }
+        else
+        {
+            qDebug("Unable to parse torrent from %s", qUtf8Printable(source));
+            m_torrentSourceCache.remove(source);
+        }
+        break;
+
+    case Net::DownloadStatus::RedirectedToMagnet:
+        if (const auto parseResult = BitTorrent::TorrentDescriptor::parse(result.magnetURI))
+        {
+            const BitTorrent::TorrentDescriptor &torrentDescr = parseResult.value();
+            const BitTorrent::InfoHash infoHash = torrentDescr.infoHash();
+            m_torrentSourceCache.insert(source, infoHash);
+
+            if (!m_torrentMetadataCache.contains(infoHash) && !BitTorrent::Session::instance()->isKnownTorrent(infoHash))
+            {
+                if (BitTorrent::Session::instance()->downloadMetadata(torrentDescr))
+                    m_torrentMetadataCache.insert(infoHash, torrentDescr);
+                else [[unlikely]]
+                    qDebug("Unable to fetch metadata for %s", qUtf8Printable(infoHash.toTorrentID().toString()));
+            }
+        }
+        else
+        {
+            qDebug("Unable to parse magnet URI %s", qUtf8Printable(result.magnetURI));
```

Comment:
Nothing to debug. Just invalid data came from outside.

Reply:
Changing this to `LogMsg`